ArmPkg/Drivers/CpuDxe/AArch64/Mmu.c

```diff
@@ -148,17 +150,34 @@ GetNextEntryAttribute (
   // Get the memory space map from GCD
   MemorySpaceMap = NULL;
   Status         = gDS->GetMemorySpaceMap (&NumberOfDescriptors, &MemorySpaceMap);
-  ASSERT_EFI_ERROR (Status);
+  // MU_CHANGE [BEGIN]: Check if the memory space map is valid
+  if (EFI_ERROR (Status)) {
+    // This function needs to define what is returned when an error occurs.
+    // Callers need to actually check the return value and add error handling.
+    ASSERT_EFI_ERROR (Status);
+    return 0;
+  }
+
+  // MU_CHANGE [END]: Check if the memory space map is valid
 
   // We cannot get more than 3-level page table
   ASSERT (TableLevel <= 3);
 
   // While the top level table might not contain TT_ENTRY_COUNT entries;
   // the subsequent ones should be filled up
   for (Index = 0; Index < EntryCount; Index++) {
-    Entry          = TableAddress[Index];
-    EntryType      = Entry & TT_TYPE_MASK;
-    EntryAttribute = Entry & TT_ATTRIBUTES_MASK;  // MU_CHANGE: Return all attributes from page table
+    Entry = TableAddress[Index];
+
+    // MU_CHANGE [BEGIN]: Convert integers safely
+    Status = SafeUint64ToUint32 (Entry, &EntryType);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "[%a] - Table address entry exceeds 32-bit.\n", __func__));
+      return 0;
+    }
+
+    EntryAttribute = EntryType & TT_ATTRIBUTES_MASK;  // MU_CHANGE: Return all attributes from page table

```

Comment:
There's another bug here in that `EntryType` (and `EntryAttribute`) are 32-bit and `TT_ATTRIBUTES_MASK` is 64-bit:

```c
#define TT_ATTRIBUTES_MASK  ((0xFFFULL << 52) | (0x3FFULL << 2))
```

This will require a deeper investigation into the way these values are getting passed around. Will follow up on that soon.

Reply:
In the 64-bit descriptor, there are 10 bits of lower attributes and 12 bits of upper attributes. This is currently applying the mask to a 32-bit descriptor value (the lower attributes).

This is assigned to `PrevEntryAttribute`:

```c
        *PrevEntryAttribute = EntryAttribute;
```

Where `PrevEntryAttribute` is also a `UINT32`:

```C
    IN OUT UINT32  *PrevEntryAttribute,
```

Which is passed to `PageAttributeToGcdAttribute()`:

```c
          SetGcdMemorySpaceAttributes (
            MemorySpaceMap,
            NumberOfDescriptors,
            *StartGcdRegion,
            (BaseAddress + (Index * TT_ADDRESS_AT_LEVEL (TableLevel))) - *StartGcdRegion,
            PageAttributeToGcdAttribute (*PrevEntryAttribute)
            );
```

Which accepts a `UINT64`:

```C
STATIC
UINT64
PageAttributeToGcdAttribute (
  IN UINT64  PageAttributes
  );
```

Which sets `EFI_MEMORY_XP` based on `TT_PXN_MASK | TT_UXN_MASK`:

```c
  // Process eXecute Never attribute
  if ((PageAttributes & (TT_PXN_MASK | TT_UXN_MASK)) != 0) {
    GcdAttributes |= EFI_MEMORY_XP;
  }
```

Where those bits are in the upper attributes:

```c
#define TT_PXN_MASK  BIT53
#define TT_UXN_MASK  BIT54                              // EL1&0
#define TT_XN_MASK   BIT54                              // EL2 / EL3
```

To attempt to get a full 64-bit value for the condition to set `EFI_MEMORY_XP` I'm going to put up another PR to use the full 64-bit value.